freqtrade/strategy/interface.py

```diff
@@ -637,35 +638,46 @@ def stop_loss_reached(self, current_rate: float, trade: Trade,
         # Initiate stoploss with open_rate. Does nothing if stoploss is already set.
         trade.adjust_stop_loss(trade.open_rate, stop_loss_value, initial=True)
 
+        # Stoploss rate and profit are calculated using open price in backtesting.
+        stop_rate = open or current_rate
+        stop_profit = trade.calc_profit_ratio(stop_rate)
+
         if self.use_custom_stoploss:
             stop_loss_value = strategy_safe_wrapper(self.custom_stoploss, default_retval=None
                                                     )(pair=trade.pair, trade=trade,
                                                       current_time=current_time,
-                                                      current_rate=current_rate,
-                                                      current_profit=current_profit)
+                                                      current_rate=stop_rate,
+                                                      current_profit=stop_profit)
             # Sanity check - error cases will return None
             if stop_loss_value:
                 # logger.info(f"{trade.pair} {stop_loss_value=} {current_profit=}")
-                trade.adjust_stop_loss(current_rate, stop_loss_value)
+                trade.adjust_stop_loss(stop_rate, stop_loss_value)
             else:
                 logger.warning("CustomStoploss function did not return valid stoploss")
 
         if self.trailing_stop:
             # trailing stoploss handling
             sl_offset = self.trailing_stop_positive_offset
-
-            # Make sure current_profit is calculated using high for backtesting.
-            high_profit = current_profit if not high else trade.calc_profit_ratio(high)
+            high_profit = trade.calc_profit_ratio(high or current_rate)
 
             # Don't update stoploss if trailing_only_offset_is_reached is true.
-            if not (self.trailing_only_offset_is_reached and high_profit < sl_offset):
+            if not self.trailing_only_offset_is_reached or high_profit > sl_offset:
                 # Specific handling for trailing_stop_positive
                 if self.trailing_stop_positive is not None and high_profit > sl_offset:
                     stop_loss_value = self.trailing_stop_positive
+                    if high is not None:
+                        # In backtesting only:
+                        if self.trailing_only_offset_is_reached:
+                            # Worst case: price reaches stop_positive_offset and dives down.
+                            stop_rate = stop_rate * (1 + abs(self.trailing_stop_positive_offset))
```

Comment:
i'm not sure i agree with this logic...

You're assuming open + positive-offset - but i don't think it's correct if you don't check if we're in a "trailing mode" already.
so you might have (with a trailing_offset of 2%, trailing with 1%):

* first candle - 1% profit (not trailing yet)
* second candle - 0.9% profit (not trailing)
* third candle - 0.5% profit (we're now trailing ... but it didn't dip down below the stoploss yet - so the trade stays open)


* 4th candle - 0.5% profit 

alternative 4th candle (case 1):
5% profit - here, you start "mid-candle" ... which might be Okish

another alternative (case 3):
-5% "full red" candle - you're now assuming "high" as your sell-price (it's limited to high, otherwise it would be higher)

In the first case - you have a stoploss > high (open + 2%)... so you sell immediately, even though the candle was "not long enough" to even cause a stoploss

the 2nd case is probably okish

In the last case - it'll be a lot higher than how it was with the previous implementation.
I think, this is fixing one particular usecase (where you have a lot of tight stoplosses) - but it's not generic enough to properly work with "wider" trailing stoploss settings ... where it can cause quite some stirr ...